File: wip/cgol.lua
#!/usr/bin/env lua

-- Constants for the grid size and density of alive cells
local WIDTH = 50
local HEIGHT = 30
local DENSITY = 0.2

-- Initialize the grid with random alive or dead cells
function initialize_grid()
    local grid = {}
    for y = 1, HEIGHT do
        grid[y] = {}
        for x = 1, WIDTH do
            grid[y][x] = math.random() < DENSITY and 1 or 0
        end
    end
    return grid
end

-- Count the number of alive neighbors around a given cell
function count_neighbors(grid, x, y)
    local count = 0
    local directions = {-1, 0, 1}

    for _, dx in ipairs(directions) do
        for _, dy in ipairs(directions) do
            if dx == 0 and dy == 0 then
                -- Skip the cell itself
                goto continue
            end

            -- Calculate neighbor coordinates with wrapping
            local nx = ((x + dx - 1) % WIDTH) + 1
            local ny = ((y + dy - 1) % HEIGHT) + 1

            if grid[ny][nx] == 1 then
                count = count + 1
            end

            ::continue::
        end
    end

    return count
end

-- Compute the next state of the grid based on the Game of Life rules
function compute_next_state(grid)
    local new_grid = initialize_grid()

    for y = 1, HEIGHT do
        for x = 1, WIDTH do
            local alive = grid[y][x]
            local neighbors = count_neighbors(grid, x, y)

            -- Apply Game of Life rules
            if (alive == 1 and (neighbors == 2 or neighbors == 3)) or
               (alive == 0 and neighbors == 3) then
                new_grid[y][x] = 1
            else
                new_grid[y][x] = 0
            end
        end
    end

    return new_grid
end

-- Print the grid to the console
function print_grid(grid)
    os.execute("clear")  -- Clear the console screen
    for y = 1, HEIGHT do
        for x = 1, WIDTH do
            if grid[y][x] == 1 then
                io.write("█")
            else
                io.write(" ")
            end
        end
        io.write("\n")
    end
end

-- Main execution function
function main()
    math.randomseed(os.time())  -- Seed the random number generator
    local grid = initialize_grid()

    while true do
        print_grid(grid)
        grid = compute_next_state(grid)
        os.execute("sleep 0.1")  -- Pause for 100 milliseconds
    end
end

-- Run the main function with error handling to catch interruptions
local status, err = xpcall(main, function(err)
    print("\nSimulation stopped due to error:", err)
end)

-- If xpcall fails, it will print an error message and exit gracefully
if not status then
    print("\nSimulation stopped.")
end

File: cgol.pas
program ConwayGameOfLife;

{$mode objfpc}{$H+}
{$codepage UTF8}

uses
  SysUtils, Crt;

const
  WIDTH = 50;
  HEIGHT = 30;
  DENSITY = 0.2;

type
  TGrid = array[0..HEIGHT-1, 0..WIDTH-1] of Boolean;

var
  grid, newGrid: TGrid;

procedure InitializeGrid(var g: TGrid);
var
  i, j: Integer;
begin
  Randomize;
  for i := 0 to HEIGHT - 1 do
    for j := 0 to WIDTH - 1 do
      g[i, j] := Random < DENSITY;
end;

function CountNeighbors(const g: TGrid; x, y: Integer): Integer;
var
  count, dx, dy, nx, ny: Integer;
begin
  count := 0;
  for dy := -1 to 1 do
  begin
    for dx := -1 to 1 do
    begin
      if not ((dx = 0) and (dy = 0)) then
      begin
        nx := (x + dx + WIDTH) mod WIDTH;
        ny := (y + dy + HEIGHT) mod HEIGHT;
        if g[ny, nx] then
          Inc(count);
      end;
    end;
  end;
  Result := count;
end;

procedure ComputeNextState(const current: TGrid; var next: TGrid);
var
  x, y, neighbors: Integer;
  isAlive: Boolean;
begin
  for y := 0 to HEIGHT - 1 do
  begin
    for x := 0 to WIDTH - 1 do
    begin
      isAlive := current[y, x];
      neighbors := CountNeighbors(current, x, y);
      if (isAlive and ((neighbors = 2) or (neighbors = 3))) or (not isAlive and (neighbors = 3)) then
        next[y, x] := True
      else
        next[y, x] := False;
    end;
  end;
end;

procedure PrintGrid(const g: TGrid);
var
  x, y: Integer;
begin
  for y := 0 to HEIGHT - 1 do
  begin
    for x := 0 to WIDTH - 1 do
    begin
      if g[y, x] then
        Write('█')
      else
        Write(' ');
    end;
    WriteLn;
  end;
end;

begin
  InitializeGrid(grid);
  while not KeyPressed do
  begin
    Write(#27'[H'#27'[2J'); // ANSI code to clear screen
    PrintGrid(grid);
    ComputeNextState(grid, newGrid);
    grid := newGrid;
    Delay(100);
  end;
end.